File: problem/F4-Network_test_maxcho/F4-network_mix/src/main.py
import socket
import sys


def log(msg):
    """Print debug log message"""
    print(f"[LOG] {msg}")


def test_redis(host, port):
    """Test Redis sidecar with AUTH command"""
    log(f"Mode: Redis test")
    log(f"Target: {host}:{port}")

    try:
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.settimeout(5)

        log("Connecting to Redis...")
        sock.connect((host, port))
        log("Connected!")

        password = "noj_secret_pass"
        command = f"AUTH {password}\r\n"
        log(f"Sending AUTH command...")
        sock.sendall(command.encode())

        response = sock.recv(1024).decode()
        log(f"Raw response: {repr(response)}")

        if response.startswith("+OK"):
            log("Redis AUTH succeeded!")
            print("good")
        else:
            log("Redis AUTH failed!")
            print("fail")

        sock.close()
    except Exception as e:
        log(f"Exception: {e}")
        print("fail")


def test_http_sidecar(host, port):
    """Test HTTP secret-server sidecar (verify_env_args_success)"""
    log(f"Mode: HTTP sidecar test")
    log(f"Target: {host}:{port}")

    try:
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.settimeout(5)

        log("Connecting...")
        sock.connect((host, port))
        log("Connected!")

        request = f"GET / HTTP/1.0\r\nHost: {host}\r\n\r\n"
        log("Sending HTTP GET request...")
        sock.sendall(request.encode())

        log("Receiving response...")
        response = b""
        while True:
            chunk = sock.recv(4096)
            if not chunk:
                break
            response += chunk

        response_str = response.decode('utf-8', errors='replace')
        log(f"Response length: {len(response_str)}")
        log(f"Content snippet: {repr(response_str[:100])}...")

        if "verify_env_args_success" in response_str:
            log("Found secret keyword!")
            print("good")
        else:
            log("Secret keyword not found!")
            print("fail")

        sock.close()
    except Exception as e:
        log(f"Exception: {e}")
        print("fail")


def test_custom_python(host, port):
    """Test custom Python environment (Hello from Server Container!)"""
    log(f"Mode: Custom Python env test")
    log(f"Target: {host}:{port}")

    try:
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.settimeout(5)

        log("Connecting...")
        sock.connect((host, port))
        log("Connected!")

        request = f"GET / HTTP/1.0\r\nHost: {host}\r\n\r\n"
        log("Sending HTTP GET request...")
        sock.sendall(request.encode())

        log("Receiving response...")
        response = b""
        while True:
            chunk = sock.recv(4096)
            if not chunk:
                break
            response += chunk

        response_str = response.decode('utf-8', errors='replace')
        log(f"Response length: {len(response_str)}")
        log(f"Content snippet: {repr(response_str[:100])}...")

        if "Hello from Server Container!" in response_str:
            log("Matched Python env signature!")
            print("good")
        else:
            log("Python env signature not found!")
            print("fail")

        sock.close()
    except Exception as e:
        log(f"Exception: {e}")
        print("fail")


def test_custom_cpp(host, port):
    """Test custom C++ environment (Hello from C++ File!)"""
    log(f"Mode: Custom C++ env test")
    log(f"Target: {host}:{port}")

    try:
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.settimeout(5)

        log("Connecting...")
        sock.connect((host, port))
        log("Connected!")

        request = f"GET / HTTP/1.0\r\nHost: {host}\r\n\r\n"
        log("Sending HTTP GET request...")
        sock.sendall(request.encode())

        log("Receiving response...")
        response = b""
        while True:
            chunk = sock.recv(4096)
            if not chunk:
                break
            response += chunk

        response_str = response.decode('utf-8', errors='replace')
        log(f"Response length: {len(response_str)}")
        log(f"Content snippet: {repr(response_str[:100])}...")

        if "Hello from C++ File!" in response_str:
            log("Matched C++ env signature!")
            print("good")
        else:
            log("C++ env signature not found!")
            print("fail")

        sock.close()
    except Exception as e:
        log(f"Exception: {e}")
        print("fail")


def test_external(host, port):
    """Test external network connectivity (IP or hostname)"""
    log(f"Mode: External network test")
    log(f"Target: {host}:{port}")

    try:
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.settimeout(3)

        log("Resolving hostname...")
        log(f"Attempting connection to {host}:{port}...")

        result = sock.connect_ex((host, port))

        if result == 0:
            log("Connection succeeded!")
            print("good")
        else:
            log(f"Connection failed with error code: {result}")
            print("fail")

        sock.close()
    except Exception as e:
        log(f"Exception: {e}")
        print("fail")


def main():
    # Read input: <type> <host> <port>
    try:
        line = sys.stdin.read().strip()
        if not line:
            log("No input provided!")
            print("fail")
            return

        parts = line.split()
        if len(parts) < 3:
            log(f"Invalid input format: {line}")
            print("fail")
            return

        test_type = parts[0]
        host = parts[1]
        port = int(parts[2])

        log(f"Input parsed: type={test_type}, host={host}, port={port}")

    except Exception as e:
        log(f"Failed to parse input: {e}")
        print("fail")
        return

    # Dispatch to appropriate test function
    if test_type == "redis":
        test_redis(host, port)
    elif test_type == "http":
        test_http_sidecar(host, port)
    elif test_type == "custom_python":
        test_custom_python(host, port)
    elif test_type == "custom_cpp":
        test_custom_cpp(host, port)
    elif test_type == "external":
        test_external(host, port)
    else:
        log(f"Unknown test type: {test_type}")
        print("fail")


if __name__ == "__main__":
    main()

File: problem/F4-Network_test_maxcho/F4-network_mix/src/main.c
#include <arpa/inet.h>
#include <netdb.h>
#include <stdio.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <unistd.h>


void log_msg(const char *msg) { printf("[LOG] %s\n", msg); }

void log_fmt(const char *fmt, const char *arg) {
  printf("[LOG] ");
  printf(fmt, arg);
  printf("\n");
}

void log_fmt_int(const char *fmt, int arg) {
  printf("[LOG] ");
  printf(fmt, arg);
  printf("\n");
}

int resolve_host(const char *host, struct sockaddr_in *server) {
  server->sin_family = AF_INET;

  if (inet_pton(AF_INET, host, &server->sin_addr) <= 0) {
    log_msg("Resolving hostname via DNS...");
    struct hostent *he = gethostbyname(host);
    if (he == NULL) {
      log_msg("DNS resolution failed!");
      return -1;
    }
    server->sin_addr = *((struct in_addr *)he->h_addr_list[0]);
  }

  char ip_str[INET_ADDRSTRLEN];
  inet_ntop(AF_INET, &server->sin_addr, ip_str, sizeof(ip_str));
  log_fmt("Resolved IP: %s", ip_str);
  return 0;
}

void test_redis(const char *host, int port) {
  log_msg("Mode: Redis test");
  log_fmt("Target host: %s", host);
  log_fmt_int("Target port: %d", port);

  struct sockaddr_in server;
  if (resolve_host(host, &server) < 0) {
    printf("fail\n");
    return;
  }
  server.sin_port = htons(port);

  int sock = socket(AF_INET, SOCK_STREAM, 0);
  if (sock == -1) {
    log_msg("Socket creation failed!");
    printf("fail\n");
    return;
  }

  log_msg("Connecting to Redis...");
  if (connect(sock, (struct sockaddr *)&server, sizeof(server)) < 0) {
    log_msg("Connection failed!");
    printf("fail\n");
    close(sock);
    return;
  }
  log_msg("Connected!");

  const char *auth_cmd = "AUTH noj_secret_pass\r\n";
  log_msg("Sending AUTH command...");
  send(sock, auth_cmd, strlen(auth_cmd), 0);

  char buffer[1024];
  memset(buffer, 0, sizeof(buffer));
  int len = recv(sock, buffer, sizeof(buffer) - 1, 0);

  if (len > 0) {
    buffer[strcspn(buffer, "\r\n")] = 0;
    log_fmt("Raw response: [%s]", buffer);

    if (strncmp(buffer, "+OK", 3) == 0) {
      log_msg("Redis AUTH succeeded!");
      printf("good\n");
    } else {
      log_msg("Redis AUTH failed!");
      printf("fail\n");
    }
  } else {
    log_msg("No response received!");
    printf("fail\n");
  }

  close(sock);
}

void test_http_sidecar(const char *host, int port) {
  log_msg("Mode: HTTP sidecar test");
  log_fmt("Target host: %s", host);
  log_fmt_int("Target port: %d", port);

  struct sockaddr_in server;
  if (resolve_host(host, &server) < 0) {
    printf("fail\n");
    return;
  }
  server.sin_port = htons(port);

  int sock = socket(AF_INET, SOCK_STREAM, 0);
  if (sock == -1) {
    log_msg("Socket creation failed!");
    printf("fail\n");
    return;
  }

  log_msg("Connecting...");
  if (connect(sock, (struct sockaddr *)&server, sizeof(server)) < 0) {
    log_msg("Connection failed!");
    printf("fail\n");
    close(sock);
    return;
  }
  log_msg("Connected!");

  char msg[256];
  snprintf(msg, sizeof(msg), "GET / HTTP/1.0\r\nHost: %s\r\n\r\n", host);
  log_msg("Sending HTTP GET request...");
  send(sock, msg, strlen(msg), 0);

  log_msg("Receiving response...");
  char buffer[4096];
  int total_len = 0;
  int len;

  while ((len = recv(sock, buffer + total_len, sizeof(buffer) - total_len - 1,
                     0)) > 0) {
    total_len += len;
    if (total_len >= sizeof(buffer) - 1)
      break;
  }
  buffer[total_len] = '\0';

  log_fmt_int("Response length: %d", total_len);

  if (strstr(buffer, "verify_env_args_success")) {
    log_msg("Found secret keyword!");
    printf("good\n");
  } else {
    log_msg("Secret keyword not found!");
    printf("fail\n");
  }

  close(sock);
}

void test_custom_python(const char *host, int port) {
  log_msg("Mode: Custom Python env test");
  log_fmt("Target host: %s", host);
  log_fmt_int("Target port: %d", port);

  struct sockaddr_in server;
  if (resolve_host(host, &server) < 0) {
    printf("fail\n");
    return;
  }
  server.sin_port = htons(port);

  int sock = socket(AF_INET, SOCK_STREAM, 0);
  if (sock == -1) {
    log_msg("Socket creation failed!");
    printf("fail\n");
    return;
  }

  log_msg("Connecting...");
  if (connect(sock, (struct sockaddr *)&server, sizeof(server)) < 0) {
    log_msg("Connection failed!");
    printf("fail\n");
    close(sock);
    return;
  }
  log_msg("Connected!");

  char msg[256];
  snprintf(msg, sizeof(msg), "GET / HTTP/1.0\r\nHost: %s\r\n\r\n", host);
  log_msg("Sending HTTP GET request...");
  send(sock, msg, strlen(msg), 0);

  log_msg("Receiving response...");
  char buffer[4096];
  int total_len = 0;
  int len;

  while ((len = recv(sock, buffer + total_len, sizeof(buffer) - total_len - 1,
                     0)) > 0) {
    total_len += len;
    if (total_len >= sizeof(buffer) - 1)
      break;
  }
  buffer[total_len] = '\0';

  log_fmt_int("Response length: %d", total_len);

  if (strstr(buffer, "Hello from Server Container!")) {
    log_msg("Matched Python env signature!");
    printf("good\n");
  } else {
    log_msg("Python env signature not found!");
    printf("fail\n");
  }

  close(sock);
}

void test_custom_cpp(const char *host, int port) {
  log_msg("Mode: Custom C++ env test");
  log_fmt("Target host: %s", host);
  log_fmt_int("Target port: %d", port);

  struct sockaddr_in server;
  if (resolve_host(host, &server) < 0) {
    printf("fail\n");
    return;
  }
  server.sin_port = htons(port);

  int sock = socket(AF_INET, SOCK_STREAM, 0);
  if (sock == -1) {
    log_msg("Socket creation failed!");
    printf("fail\n");
    return;
  }

  log_msg("Connecting...");
  if (connect(sock, (struct sockaddr *)&server, sizeof(server)) < 0) {
    log_msg("Connection failed!");
    printf("fail\n");
    close(sock);
    return;
  }
  log_msg("Connected!");

  char msg[256];
  snprintf(msg, sizeof(msg), "GET / HTTP/1.0\r\nHost: %s\r\n\r\n", host);
  log_msg("Sending HTTP GET request...");
  send(sock, msg, strlen(msg), 0);

  log_msg("Receiving response...");
  char buffer[4096];
  int total_len = 0;
  int len;

  while ((len = recv(sock, buffer + total_len, sizeof(buffer) - total_len - 1,
                     0)) > 0) {
    total_len += len;
    if (total_len >= sizeof(buffer) - 1)
      break;
  }
  buffer[total_len] = '\0';

  log_fmt_int("Response length: %d", total_len);

  if (strstr(buffer, "Hello from C++ File!")) {
    log_msg("Matched C++ env signature!");
    printf("good\n");
  } else {
    log_msg("C++ env signature not found!");
    printf("fail\n");
  }

  close(sock);
}

void test_external(const char *host, int port) {
  log_msg("Mode: External network test");
  log_fmt("Target host: %s", host);
  log_fmt_int("Target port: %d", port);

  struct sockaddr_in server;
  if (resolve_host(host, &server) < 0) {
    printf("fail\n");
    return;
  }
  server.sin_port = htons(port);

  int sock = socket(AF_INET, SOCK_STREAM, 0);
  if (sock == -1) {
    log_msg("Socket creation failed!");
    printf("fail\n");
    return;
  }

  struct timeval timeout;
  timeout.tv_sec = 3;
  timeout.tv_usec = 0;
  setsockopt(sock, SOL_SOCKET, SO_SNDTIMEO, (const char *)&timeout,
             sizeof(timeout));

  log_msg("Attempting connection...");
  if (connect(sock, (struct sockaddr *)&server, sizeof(server)) < 0) {
    log_msg("Connection failed!");
    printf("fail\n");
  } else {
    log_msg("Connection succeeded!");
    printf("good\n");
  }

  close(sock);
}

int main() {
  char type[32];
  char host[128];
  int port;

  if (scanf("%31s %127s %d", type, host, &port) != 3) {
    log_msg("Invalid input format!");
    printf("fail\n");
    return 0;
  }

  log_fmt("Input type: %s", type);
  log_fmt("Input host: %s", host);
  log_fmt_int("Input port: %d", port);

  if (strcmp(type, "redis") == 0) {
    test_redis(host, port);
  } else if (strcmp(type, "http") == 0) {
    test_http_sidecar(host, port);
  } else if (strcmp(type, "custom_python") == 0) {
    test_custom_python(host, port);
  } else if (strcmp(type, "custom_cpp") == 0) {
    test_custom_cpp(host, port);
  } else if (strcmp(type, "external") == 0) {
    test_external(host, port);
  } else {
    log_fmt("Unknown test type: %s", type);
    printf("fail\n");
  }

  return 0;
}